### ISSUE TITLE
GC issue preventing SR deletion with LVHDSR driver

### ISSUE THREAD

**Issue by stormi** [xapi-project/sm#583]
We had an issue in our automated storage testing, where we create a LVHDSR, import a small VM, move it around in the pool (cold migration to second pool host then back to master, then live migration to second host and back to master).

What fails is our teardown phase when, after deleting the only VM on the SR, we try to delete the SR (`xe sr-destroy`). It fails because the SR is not empty.

The SR delete function does trigger a GC run, but it leaves one VDI behind:

```
Mar 22 19:22:31 r620-q2 SMGC: [29353] SR 4659 ('LVM-local-SR') (6 VDIs in 5 VHD trees):
Mar 22 19:22:31 r620-q2 SMGC: [29353]         *8341e744[VHD](2.000G//2.012G|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]             *53b46c10[VHD](2.000G//40.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]         *e7657971[VHD](2.000G//2.012G|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]         *43b9baf6[VHD](2.000G//8.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]         *080f8024[VHD](2.000G//8.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]         *4f3fa3ba[VHD](2.000G//172.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]
Mar 22 19:22:31 r620-q2 SMGC: [29353] Found 5 VDIs for deletion:
Mar 22 19:22:31 r620-q2 SMGC: [29353]   *53b46c10[VHD](2.000G//40.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]   *e7657971[VHD](2.000G//2.012G|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]   *43b9baf6[VHD](2.000G//8.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]   *080f8024[VHD](2.000G//8.000M|n)
Mar 22 19:22:31 r620-q2 SMGC: [29353]   *4f3fa3ba[VHD](2.000G//172.000M|n)
```

The parent of 53b46c10 is not selected for deletion.

If we re-attach the SR and force a new garbage collection (or wait for it to run once again, but we can't wait for so long in automated tests), then the last VDI is GCed too:

```
Mar 22 19:22:43 r620-q2 SMGC: [30330] SR 4659 ('LVM-local-SR') (1 VDIs in 1 VHD trees):
Mar 22 19:22:43 r620-q2 SMGC: [30330]         *8341e744[VHD](2.000G//2.012G|n)
Mar 22 19:22:43 r620-q2 SMGC: [30330]
Mar 22 19:22:43 r620-q2 SMGC: [30330] Found 1 VDIs for deletion:
Mar 22 19:22:43 r620-q2 SMGC: [30330]   *8341e744[VHD](2.000G//2.012G|n)
```

And then we can delete the SR properly.

Is this the expected behaviour of garbage collection, not removing the whole VHD tree even if the whole VHD is not useful anymore? If so, should it run several times when force-triggered by a SR delete, to ensure all removable VHDs are properly removed and SR delete doesn't fail?

**Comment by MarkSymsCtx** [comment]
Yes this is expected behaviour, `sr destroy` rightly refuses to destroy an SR with data on it as if that's done accidentally data loss will occur. So there must be no VDIs, even hidden ones when the `sr destroy` is issued. As this is happening in your automated testing you can use this command to check that the SR garbage collection is completed before unplugging it.
```
/opt/xensource/sm/cleanup -q <sr uuid>
```
Which will return "Currently running: True|False"

The GC process should run until there is nothing left to do so as soon as the above returns False it should be possible to issue the `sr destroy`

**Comment by stormi** [comment]
Thanks for the answer. My point was that, as the driver's `delete` function already does trigger a forced GC before attempting to delete the SR, I was hoping that this GC would be sufficient. If it's a requirement that the SR is already empty before running `sr_destroy`, then I'm not sure what's the point of the driver's `delete` function triggering a GC before it attempts the deletion. 
Is it just a way to favour SR deletion success in cases when the only remaining VDIs are leafs, but without ensuring this will work when it's not the case?

**Comment by MarkSymsCtx** [comment]
Are you sure that GC process completed? If it did then it's a bug as the GC should run to completion, if it didn't and you unplugged the SR before it completed then that's the answer.

**Comment by stormi** [comment]
I'm unsure how the GC works so I'll try to lay out the successive steps for you to judge :)

* initial state: all VMs and VDIs deleted, but GC still needs to run to collect the hidden ones
* *without waiting*, I run `xe sr-destroy`
* This calls the `delete` function from `LVHDSR.py`, which, before anything, runs `cleanup.gc_force` (https://github.com/xapi-project/sm/blob/317a64e748d8b119d88231e0113199ba9681e14b/drivers/LVHDSR.py#L520). This selects 5 out of 6 VDIs for deletion (log extract in my first message), but the run completes without removing the sixth VDI.
* The `delete` function keeps executing, until it fails because of the last VDI not being removed by the GC run.

**Comment by MarkSymsCtx** [comment]
For the purposes of your automation I would suggest adding the following immediately before the `xe sr-destroy`

- xe sr-scan <sr uuid>
- /opt/xensource/cleanup -q <sr uuid> (until Running = False)

**Comment by stormi** [comment]
Thanks, I'll try this. Any idea why there's a forced GC in `delete`?

**Comment by stormi** [comment]
So I implemented it and it works, but it may add a significant delay on the test runs. In the example below, the script had to wait for over 3 minutes for the cleanup to complete. 

At 11:17:46 I ran sr_scan, which deleted a VDI:
```
Mar 24 11:17:47 r620-q2 SM: [18123] VDIs missing from disk: set(['c1d657be-81d5-4047-80ff-fa33078703ac'])
Mar 24 11:17:47 r620-q2 SM: [18123] Forgetting VDI with location=c1d657be-81d5-4047-80ff-fa33078703ac uuid=c1d657be-81d5-4047-80ff-fa33078703ac
Mar 24 11:17:47 r620-q2 SM: [18123] Deleting vdi: c1d657be-81d5-4047-80ff-fa33078703ac
```

Then it started GC:
```
Mar 24 11:17:47 r620-q2 SM: [18123] Kicking GC
```

But there already was an instance running so it died:
```
Mar 24 11:17:47 r620-q2 SMGC: [18175] Another GC instance already active, exiting
```

And it's only at 11:21:04 that the other GC instance decided to leave its "quiet period":
```
Mar 24 11:21:04 r620-q2 SMGC: [11620]   Child process completed successfully
Mar 24 11:21:04 r620-q2 SMGC: [11620] GC active, quiet period ended
```

As I don't really care about the SR that I'm going to destroy in the context of test automation, I may try more aggressive approaches such as running sr_destroy several times in a row if the reason it failed is SR_NOT_EMPTY, or force running the GC after the scan to avoid having to wait for the end of the quiet period :thinking:
